### DISCUSSION TITLE
Migration for Multi tenancy

### DISCUSSION THREAD

**Discussion by guschinaglia** [prisma/prisma#24784] [Q&A]
### Question

I'm creating a Multitenant service using NestJS with Prisma and CockroachDB. 
When user register, it does create a new schema but doesn't create the tables for the new Tenant.
This generates a few problemas, like initial load etc.

There is any way without having to write down every SQL statement to create the tables using Prisma programatically ?

**Comment by ludralph** [comment]
Hi @guschinaglia 👋 

Thank you for raising this question. Can you please answer the following questions so that we can better help you?

- How are you currently creating the new schema for each tenant? 
- Can you share the Prisma schema you are using? This will help in understanding the models and relationships you have defined.
- How are you managing database connections for different tenants? Are you using a single connection string or multiple connection strings for different schemas?
- Are you encountering any specific error message when trying to create tables for the new tenant? If so, can you please share it?

**Comment by guschinaglia** [comment]
Hello @ludralph :D Thank you for the help!

Sure i can give you some infos about it:

First off all, i'm using NestJS.

- When the validations passes after the user registration, i just simply trigger a create schema for it:
`   await this.prismaService.$executeRawUnsafe(
      CREATE SCHEMA IF NOT EXISTS "${schemaName}"
    );
`

- I will post an example pretty similar to what i have:

![image](https://github.com/user-attachments/assets/3ffb5db9-9647-4cad-92b3-d2a2ed7691cc)

- When the tenant is passed as a parameter, i select his correct schema:

![image](https://github.com/user-attachments/assets/e4d4b9cd-cd96-40f7-b6fd-a8b39ccf3aeb)

![image](https://github.com/user-attachments/assets/666243e7-8bea-4668-b961-aa710e9695e2)

- The issue happens when i try to pre-populate the "Posts" for example using a JS function, and have errors since the Table doesn't exists in the Schema.

I Want to pre-create all tables after i do the first step which is the "create schema".
With this, i can guarantee that all the tables are created in the new schema.

**Reply by ludralph** [reply]
Hi @guschinaglia 

Your current issue is that after creating a new schema for a tenant, the tables within that schema are not automatically created, causing errors when you try to pre-populate data.

To solve this issue and create all the necessary tables in the new schema after creating it, you can use Prisma's `db push` command programmatically. This approach will create the tables based on your Prisma schema without generating migrations. Here's how you can implement this:

First, install the @prisma/internals package:
```
npm install @prisma/internals
```
Create a utility function to push the schema to the new tenant's schema:
```ts
import { execSync } from 'child_process';
import { getDMMF, getSchemaPath } from '@prisma/internals';

async function pushSchemaToTenant(schemaName: string) {
  const schemaPath = getSchemaPath();
  const dmmf = await getDMMF({ datamodelPath: schemaPath });

  // Temporarily modify the schema to use the tenant's schema
  const originalSchema = dmmf.datamodel.raw;
  const modifiedSchema = originalSchema.replace(
    /provider\s*=\s*"postgresql"/,
    `provider = "postgresql"\n  schemas = ["${schemaName}"]`
  );

  // Write the modified schema to a temporary file
  const tempSchemaPath = `${schemaPath}.temp`;
  fs.writeFileSync(tempSchemaPath, modifiedSchema);

  try {
    // Push the schema to the tenant's schema
    execSync(`npx prisma db push --schema ${tempSchemaPath} --accept-data-loss`, { stdio: 'inherit' });
  } finally {
    // Clean up the temporary schema file
    fs.unlinkSync(tempSchemaPath);
  }
}
```
Modify your tenant creation logic to use this function:
```ts
async createTenant(schemaName: string) {
  // Create the schema
  await this.prismaService.$executeRawUnsafe(
    `CREATE SCHEMA IF NOT EXISTS "${schemaName}"`
  );

  // Push the Prisma schema to the new tenant's schema
  await pushSchemaToTenant(schemaName);

  // Now you can pre-populate data if needed
  await this.prePolulateTenantData(schemaName);
}
```
This approach will create all the tables defined in your Prisma schema within the new tenant's schema. After this, you should be able to pre-populate data without encountering "table does not exist" errors.

**Comment by ludralph** [comment]
Hi there,

To keep our discussions organized and focused on the most relevant topics, we’re reviewing and tidying up our backlog. As part of this process, we’re closing discussions that haven’t had any recent activity and appear to be outdated.

If this discussion is still important to you or unresolved, we’d love to hear from you! Feel free to reopen it or start a new one with updated details.

For more details about our priorities and vision for the future of Prisma ORM, check out our latest blog post: https://www.prisma.io/blog/prisma-orm-manifesto.

Thank you for your understanding and being part of the community!;
